File: test_backend_access.js
#!/usr/bin/env node
/**
 * Test script for direct Monkey Coder backend API access using Node.js.
 * 
 * This script demonstrates how to integrate with the Monkey Coder backend
 * from external Node.js projects.
 */

class MonkeyCoderBackendClient {
    constructor(baseUrl = 'http://localhost:8000', apiKey = null) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.apiKey = apiKey;
    }

    async makeRequest(endpoint, options = {}) {
        const url = `${this.baseUrl}${endpoint}`;
        const headers = {
            'Content-Type': 'application/json',
            ...options.headers
        };

        if (this.apiKey) {
            headers.Authorization = `Bearer ${this.apiKey}`;
        }

        const response = await fetch(url, {
            ...options,
            headers
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        return await response.json();
    }

    async createDevApiKey() {
        return await this.makeRequest('/api/v1/auth/keys/dev', {
            method: 'POST'
        });
    }

    async checkHealth() {
        return await this.makeRequest('/health');
    }

    async getAuthStatus() {
        if (!this.apiKey) {
            throw new Error('API key required for authentication check');
        }
        return await this.makeRequest('/api/v1/auth/status');
    }

    async executeTask(prompt, taskType = 'code_generation', persona = 'developer', options = {}) {
        if (!this.apiKey) {
            throw new Error('API key required for task execution');
        }

        const taskData = {
            task_id: `api-task-${Math.floor(Math.random() * 1000000)}`,
            task_type: taskType,
            prompt: prompt,
            context: {
                user_id: 'api-user',
                session_id: `session-${Math.floor(Math.random() * 1000)}`,
                environment: 'nodejs-backend-access',
                ...options.context
            },
            persona_config: {
                persona: persona,
                ...options.persona_config
            },
            ...options
        };

        return await this.makeRequest('/api/v1/execute', {
            method: 'POST',
            body: JSON.stringify(taskData)
        });
    }

    async getCapabilities() {
        return await this.makeRequest('/api/v1/capabilities');
    }
}

async function main() {
    console.log('🐒 Monkey Coder Backend API Access Test (Node.js)\n');

    const client = new MonkeyCoderBackendClient();

    // Test 1: Health check
    console.log('1. Testing backend health...');
    try {
        const health = await client.checkHealth();
        console.log(`   ✓ Backend is ${health.status}`);
        console.log(`   ✓ Version: ${health.version}`);
    } catch (error) {
        console.log(`   ✗ Health check failed: ${error.message}`);
        return;
    }

    // Test 2: Create API key
    console.log('\n2. Creating development API key...');
    let apiKey;
    try {
        const keyData = await client.createDevApiKey();
        apiKey = keyData.key;
        console.log(`   ✓ Created API key: ${apiKey.substring(0, 20)}...`);
        
        // Update client with new key
        client.apiKey = apiKey;
    } catch (error) {
        console.log(`   ✗ API key creation failed: ${error.message}`);
        return;
    }

    // Test 3: Authentication status
    console.log('\n3. Testing authentication...');
    try {
        const authStatus = await client.getAuthStatus();
        if (authStatus.authenticated) {
            const user = authStatus.user || {};
            console.log(`   ✓ Authenticated as: ${user.name || 'Unknown'}`);
            console.log(`   ✓ Subscription tier: ${user.subscription_tier || 'Unknown'}`);
            console.log(`   ✓ Credits: $${((user.credits || 0) / 100).toFixed(2)}`);
        } else {
            console.log('   ✗ Authentication failed');
            return;
        }
    } catch (error) {
        console.log(`   ✗ Authentication check failed: ${error.message}`);
        return;
    }

    // Test 4: Get capabilities
    console.log('\n4. Getting system capabilities...');
    try {
        const capabilities = await client.getCapabilities();
        console.log(`   ✓ Available features: ${(capabilities.features || []).length}`);
        console.log(`   ✓ Available models: ${(capabilities.models || []).length}`);
    } catch (error) {
        console.log(`   ✗ Capabilities check failed: ${error.message}`);
    }

    // Test 5: Execute a simple task
    console.log('\n5. Testing task execution...');
    try {
        const result = await client.executeTask(
            'Create a simple Hello World function in JavaScript',
            'code_generation'
        );

        const taskId = result.task_id || 'Unknown';
        const status = result.status || 'Unknown';
        const confidence = result.confidence || 0;

        console.log(`   ✓ Task executed: ${taskId}`);
        console.log(`   ✓ Status: ${status}`);
        console.log(`   ✓ Confidence: ${confidence}%`);

        // Show first 100 chars of result if available
        if (result.result) {
            const resultText = JSON.stringify(result.result).substring(0, 100);
            console.log(`   ✓ Result preview: ${resultText}...`);
        }
    } catch (error) {
        console.log(`   ✗ Task execution failed: ${error.message}`);
    }

    console.log('\n✅ Backend API access test completed!');
    console.log(`🔑 API Key: ${apiKey}`);
    console.log('📚 Full documentation: BACKEND_API_ACCESS.md');
}

// Run the test
main().catch(console.error);

File: tests/integration/test_backend_access.py
#!/usr/bin/env python3
"""
Test script for direct Monkey Coder backend API access.

This script demonstrates how to integrate with the Monkey Coder backend
from external Python projects.
"""

import requests
import json
import os
from typing import Dict, Any, Optional

class MonkeyCoderBackendClient:
    """Client for direct backend API access."""
    
    def __init__(self, base_url: str = "http://localhost:8000", api_key: Optional[str] = None):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.session = requests.Session()
        
        if api_key:
            self.session.headers.update({
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json"
            })
    
    def create_dev_api_key(self) -> Dict[str, Any]:
        """Create a development API key (no authentication required)."""
        response = self.session.post(f"{self.base_url}/api/v1/auth/keys/dev")
        response.raise_for_status()
        return response.json()
    
    def check_health(self) -> Dict[str, Any]:
        """Check backend health status."""
        response = self.session.get(f"{self.base_url}/health")
        response.raise_for_status()
        return response.json()
    
    def get_auth_status(self) -> Dict[str, Any]:
        """Check authentication status."""
        if not self.api_key:
            raise ValueError("API key required for authentication check")
        
        response = self.session.get(f"{self.base_url}/api/v1/auth/status")
        response.raise_for_status()
        return response.json()
    
    def execute_task(self, 
                    prompt: str, 
                    task_type: str = "code_generation",
                    persona: str = "developer",
                    **kwargs) -> Dict[str, Any]:
        """Execute an AI task."""
        if not self.api_key:
            raise ValueError("API key required for task execution")
        
        task_data = {
            "task_id": f"api-task-{hash(prompt) % 1000000}",
            "task_type": task_type,
            "prompt": prompt,
            "context": {
                "user_id": "api-user",
                "session_id": f"session-{hash(prompt) % 1000}",
                "environment": "direct-backend-access",
                **kwargs.get("context", {})
            },
            "persona_config": {
                "persona": persona,
                **kwargs.get("persona_config", {})
            },
            **{k: v for k, v in kwargs.items() if k not in ["context", "persona_config"]}
        }
        
        response = self.session.post(f"{self.base_url}/api/v1/execute", json=task_data)
        response.raise_for_status()
        return response.json()
    
    def get_capabilities(self) -> Dict[str, Any]:
        """Get system capabilities."""
        response = self.session.get(f"{self.base_url}/api/v1/capabilities")
        response.raise_for_status()
        return response.json()

def main():
    """Demonstrate backend API access."""
    print("🐒 Monkey Coder Backend API Access Test\n")
    
    # Initialize client
    client = MonkeyCoderBackendClient()
    
    # Test 1: Health check
    print("1. Testing backend health...")
    try:
        health = client.check_health()
        print(f"   ✓ Backend is {health['status']}")
        print(f"   ✓ Version: {health['version']}")
    except Exception as e:
        print(f"   ✗ Health check failed: {e}")
        return
    
    # Test 2: Create API key
    print("\n2. Creating development API key...")
    try:
        key_data = client.create_dev_api_key()
        api_key = key_data['key']
        print(f"   ✓ Created API key: {api_key[:20]}...")
        
        # Update client with new key
        client.api_key = api_key
        client.session.headers.update({"Authorization": f"Bearer {api_key}"})
        
    except Exception as e:
        print(f"   ✗ API key creation failed: {e}")
        return
    
    # Test 3: Authentication status
    print("\n3. Testing authentication...")
    try:
        auth_status = client.get_auth_status()
        if auth_status.get('authenticated'):
            user = auth_status.get('user', {})
            print(f"   ✓ Authenticated as: {user.get('name', 'Unknown')}")
            print(f"   ✓ Subscription tier: {user.get('subscription_tier', 'Unknown')}")
            print(f"   ✓ Credits: ${(user.get('credits', 0) / 100):.2f}")
        else:
            print("   ✗ Authentication failed")
            return
    except Exception as e:
        print(f"   ✗ Authentication check failed: {e}")
        return
    
    # Test 4: Get capabilities
    print("\n4. Getting system capabilities...")
    try:
        capabilities = client.get_capabilities()
        print(f"   ✓ Available features: {len(capabilities.get('features', []))}")
        print(f"   ✓ Available models: {len(capabilities.get('models', []))}")
    except Exception as e:
        print(f"   ✗ Capabilities check failed: {e}")
    
    # Test 5: Execute a simple task
    print("\n5. Testing task execution...")
    try:
        result = client.execute_task(
            prompt="Create a simple Hello World function in Python",
            task_type="code_generation"
        )
        
        task_id = result.get('task_id', 'Unknown')
        status = result.get('status', 'Unknown')
        confidence = result.get('confidence', 0)
        
        print(f"   ✓ Task executed: {task_id}")
        print(f"   ✓ Status: {status}")
        print(f"   ✓ Confidence: {confidence}%")
        
        # Show first 100 chars of result if available
        if 'result' in result:
            result_text = str(result['result'])[:100]
            print(f"   ✓ Result preview: {result_text}...")
            
    except Exception as e:
        print(f"   ✗ Task execution failed: {e}")
    
    print(f"\n✅ Backend API access test completed!")
    print(f"🔑 API Key: {api_key}")
    print(f"📚 Full documentation: BACKEND_API_ACCESS.md")

if __name__ == "__main__":
    main()